### PR TITLE
Greg/validator/setup

### PR DESCRIPTION
**Pull request by GregTheGreek**
This sets up the initial plumbing for the validator client and should not include the full feature set. 

Checklist
- [x] Stubbed RPC service
- [x] Configuration of keystores
- [x] Check if validator is active

Things that won't be included:
- [ ] Validator Responsibilities (block processing, attestation processing)
- [ ] Performing eth1 deposit

### CODE REVIEW

File: validator/constants.ts [hunk 1]
```diff
@@ -0,0 +1,6 @@
+export const ETH1_FOLLOW_DISTANCE = 2**10; //1024 blocks ~4 hours
+export const KEYSTORE_DIR = "./keystores";
+
+
+// TEMP PLEASE REMOVE
```
Comment (wemeetagain) [inline validator/constants.ts]: Yeah we can get this from the other constants

File: validator/types.ts [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+  // Password for public keystore
+  publicKeystorePassword: string;
+  // This key can be kept cold and is used upon exit
+  withdrawalKeystore: string;
+  // Password for withdrawal keystore
+  withdrawalKeystorePassword: string;
+  // Connection string to RPC
+  rpcUrl: string;
+}
+export const ValidatorCtx = {
```
Comment (wemeetagain) [inline validator/types.ts]: We probably don't need this const. We only need const types for things that we're planning on ssz-ing.

File: validator/index.ts [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+  async validateKeystores() {
+    logger.info("Unlocking wallets...");
+
+    // Attempt to unlock public wallet
+    const [err1, publicWallet] = await to<ethers.Wallet>(unlockWallet(this.ctx.publicKeystore, this.ctx.publicKeystorePassword, "public wallet"));
+    if (err1) throw new Error("Public wallet could not be unlocked!");
+    this.publicWallet = publicWallet;
+
+    // Attempt to unlock withdrawal wallet
+    let [err2, withdrawalWallet] = await to<ethers.Wallet>(unlockWallet(this.ctx.withdrawalKeystore, this.ctx.withdrawalKeystorePassword, "withrawal wallet"));
```
Comment (wemeetagain) [inline validator/index.ts]: I think we may not want to do this. The spec mentions that the withdrawal wallet can stay in "cold storage".

File: validator/index.ts [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+
+    logger.info("Wallets successfully unlocked!");
+  }
+
+  async setup() {
+    await this.validateKeystores();
+    this.setupRPC();
+  }
+
+  start(): void {
```
Comment (wemeetagain) [inline validator/index.ts]: lets make this an async function.

then we can `await` the setup below.

File: validator/index.ts [hunk 5]
```diff
@@ -0,0 +1,48 @@
+import blgr from 'blgr';
+import ethers from "ethers";
+import to from 'await-to-js';
+import {ValidatorCtx} from "./types";
+import {unlockWallet} from "./utils/wallet";
+const logger = blgr.logger('debug');
```
Comment (wemeetagain) [inline validator/index.ts]: I think we should pass a logger instance into the object below, and then use that instance everywhere.
That way we can configure it once and have the setting used everywhere.
maybe something like
```javascript
constructor(..., logger) {
  this.logger = logger;
}

/// later
this.logger.info("Unlocking wallets...");
```

**Review by wemeetagain — COMMENTED** [review]
Looking good :)

File: validator/index.ts (hunk 3 above)
Comment (GregTheGreek) [reply validator/index.ts]: hmm good point. I'm thinking we can validate that the wallet is correct? Perhaps I destroy the object in mem?

File: validator/types.ts (hunk 2 above)
Comment (GregTheGreek) [reply validator/types.ts]: that makes a  lot of sense

File: validator/index.ts (hunk 3 above)
Comment (wemeetagain) [reply validator/index.ts]: I'm thinking most people would probably have this key offline until they want to withdraw. We probably should only import/validate/use the wallet once a withdrawal is triggered.

eg:
```javascript
withdraw(keystore, password, etc... ) {
  // create withdrawal wallet
  ...
  // do withdraw
  ...
  // destroy withdrawal wallet
  ...
}
```

Comment (GregTheGreek) [reply validator/index.ts]: Hmm ok. That works. I'm pretty sure we need to know the withdrawal address in advance let me double check

Comment (GregTheGreek) [reply validator/index.ts]: Yeah deposit info Requirea withdrawal credentials... 

My concern is about safety ensuring they used the correct public address

Comment (wemeetagain) [reply validator/index.ts]: I think the DepositInput stores a hash of the public key of a 'withdrawal account'.
That sort of thing will definitely need to be checked, but I think its more appropriate when the user initially inputs their deposit on the eth1 chain.

At this point, the deposit on the eth1 chain has presumably already cleared, and if they didn't put the right withdrawal key, or lost it somewhere along the way, there's nothing we can do for them.

Comment (GregTheGreek) [reply validator/index.ts]: Hmmmmmmmmmmmmmmm ok Im good with that assumption. I'll clean this up and redo the withrawal steps

**Comment by GregTheGreek** [comment]
@wemeetagain check my latest [commit](https://github.com/ChainSafeSystems/lodestar/pull/137/commits/61f3d1b73a0bc896e7649e857610b48333c0dd72) let me know what you think

File: validator/utils.ts [hunk 6]
```diff
@@ -0,0 +1,18 @@
+/**
+ * Function that loops over an async function until it returns a truthy value.
+ * @param {Function} fn - Async function to itterate over
+ * @param {object} params - any params for the function.
+ * @param {string} inMsg - Log message to be displayed for each loop
+ * @param {string} outMsg - Log message to be displayed upon returning a value
+ * @returns {T} - returns generic value based fn return.
+ */
+export async function loop<T>(fn: Function, params: Array, inMsg: string, outMsg: string): T {
```
Comment (wemeetagain) [inline validator/utils.ts]: Is this used anywhere?

**Review by wemeetagain — APPROVED** [review]

File: validator/utils.ts (hunk 6 above)
Comment (GregTheGreek) [reply validator/utils.ts]: Ah shoot I meant to finish its accompanying test. Its suppose to be an interim solution to `polling` essentially replace the while loops everywhere.,.. I came up with it on the plane lol, thoughts?

**Pull request merged**